### ISSUE TITLE
Linux FreeTDS and unixODBC Configuration

### ISSUE THREAD

**Issue by BHare1985** [gaspardle/mapnik-mssql#10]
OK so it's  been a journey, started off with https://github.com/gaspardle/mapnik-mssql/issues/9 and now I have finally come to the conclusion/understanding that I need `unixODBC` to compile the plugin, and `unixODBC` and `FreeTDS` to run the plugin.

I am getting the plugin to run on Amazon Linux, the default AMI with NodeJS beanstalk (using node-mapnik). To start off installing unixODBC and FreeTDS I followed https://www.caseylabs.com/how-to-install-unixodbc-and-freetds-on-centos-6-5-for-microsoft-sql-server-connections/

Things I have noticied with FreeTDS 

the DRIVER string that is sent with the plugin's connection string is `DRIVER={SQL Server Native Client 11.0}`. Probably because it can be seemless with SQL Server on windows, but that means you have to consider the driver name when creating an entry in FreeTDS.

 If FreeTDS if you run `osql -S server-dsn -U user -P password` (debugging helper) where server-dsn is defined in `odbc.ini` as

```
[server-dsn]
Driver = SQL Server Native Client 11.0
Description = Microsoft SQL
Server = 1.2.3.4
Port = 1433
Database = TestDatabase
TDS_Version = 8.0
```

then in it will attempt to look for a driver `SQLServerNativeClient11.0` in `odcbsinst.ini` rather than `SQL Server Native Client 11.0`. So all the internal debugging and usage of DSN rely on having no spaces in the driver name, but if you only create `` then the plugin won't find the correct driver because when doing a DSN-less config (see below) it will consider spaces.  

TLDR; You need to define both [SQL Server Native Client 11.0] and [SQLServerNativeClient11.0] in `odcbsinst.ini` for it work both ways.

---

The second gotcha is that `TDS_Version = 7.x` is required for SQL Server and 7.3 is the best (latest). By default it is set to 5.0. You can setup a DSN like I show above that manually sets it, but in order to use the dsn you must either specify `DSN=` or `SERVERNAME=` in connection string, whereas this plugin uses `SERVER=` which effectively is a dsn-less config ( http://freetds.schemamania.org/userguide/dsnless.htm )

You do have the option to set default TDS version inside `freetds.conf` but that won't apply when doing a DSN-less config with SERVER (See Example 4-3 in link above). The only way to have it do version 7.3 is to actually specify 7.3 when you configure the compiling with `--with-tdsver`, if you don't specify it the default is 5.0 and will result in `[unixODBC][FreeTDS][SQL Server]Unable to connect: Adaptive Server is unavailable or does not exist (2) SQLState: 08001 NativeError: 0 Message: [unixODBC][FreeTDS][SQL Server]Unable to connect to data source`

I compiled FreeTDS with `--with-tdsver=8.0` and it's still defaulting to 5.0 which stumped me. Everywhere online I looked I saw people use 8.0, so I had no idea why it wasn't working. I then found http://www.freetds.org/userguide/choosingtdsprotocol.htm so 8.0 is actually 7.1, and looking at the source code for the configure script if you supply 8.0, it will default to 5.0 and not 7.1 (lol). So use `--with-tdsver=7.3` when compiling.

Perhaps worth considering adding `TDS_Version=7.3;` in the connection string for those who try using FreeTDS in the future, or better yet looking into avoiding the whole driver specifying and all this extra steps for Linux.

**Comment by BHare1985** [comment]
Still getting `Unable to connect to data source` even though default version is now `7.3`. Using the FreeTDS `tsql` `isql` and `osql` debug commands all work correctly, so I am giving up for now. Suggestions welcomed

**Comment by BHare1985** [comment]
I had enabled logging both for ODBC and FreeTDS. Oddly though, the plugin doesnt trigger any events in the logs so I had to download `php-odbc` and `php-cli` and run a php script to test out the connection string. 

Using PHP  I got the same error, which wasn't helpful but it did populate the logs, checking the freetds log I found `login.c:436:invalid port number` which was odd, because it should assume port 1433 (I never supplied it in the config/connectionstring). Turns out that was the problem. I had to specify it was port 1433 because freetds for some reason does not make the assumption.

**Comment by BHare1985** [comment]
Spoke too soon, seems like I am getting geometry OK but additional fields are defaulting to zero. It is saying my fields additional fields are 0, when they are non-zero when I ran the SQL locally and through tsql.

Using TDS version 7.3 when I do 

```
[root@ip-172-31-38-225 nodejs]# osql -S connect -U ttds -P ttds
checking shared odbc libraries linked to isql for default directories...
strings: '': No such file
        trying /tmp/sql ... no
        trying /tmp/sql ... no
        trying /etc ... OK
checking odbc.ini files
        reading /root/.odbc.ini
[connect] not found in /root/.odbc.ini
        reading /etc/odbc.ini
[connect] found in /etc/odbc.ini
found this section:
        [connect]
        Driver          = SQL Server Native Client 11.0
        Description             = Microsoft SQL
        Server          = 68.43.177.158
        Port            = 1433
        Database                = TTDS
        TDS_Version = 7.3
looking for driver for DSN [connect] in /etc/odbc.ini
  found driver line: "  Driver          = SQL Server Native Client 11.0"
  driver "SQLServerNativeClient11.0" found for [connect] in odbc.ini
found driver named "SQLServerNativeClient11.0"
"SQLServerNativeClient11.0" is not an executable file
looking for entry named [SQLServerNativeClient11.0] in /etc/odbcinst.ini
  found driver line: "  Driver          = /usr/lib//libtdsodbc.so"
  found driver /usr/lib//libtdsodbc.so for [SQLServerNativeClient11.0] in odbcinst.ini
/usr/lib//libtdsodbc.so is an executable file
"Server" found, not using freetds.conf
Server is "68.43.177.158"
looking up hostname for ip address 68.43.177.158

Configuration looks OK.  Connection details:

                   DSN: connect
              odbc.ini: /etc/odbc.ini
                Driver: /usr/lib//libtdsodbc.so
       Server hostname: c-68-43-177-158.hsd1.mi.comcast.net
               Address: 68.43.177.158

Attempting connection as ttds ...
+ isql connect ttds ttds -v
+---------------------------------------+
| Connected!                            |
|                                       |
| sql-statement                         |
| help [tablename]                      |
| quit                                  |
|                                       |
+---------------------------------------+
SQL> SELECT "Geometry" AS geom,"StyleIndex" FROM (SELECT Geometry,LinkDir,StyleIndex,IsOneWay FROM MAPNIK WHERE SessionId = 69) as cdbq WHERE "Geometry".STIntersects(geometry::STGeomFromText('POLYGON((-92.81250000000001 29.53522956294844,-75.93749999999999 29.53522956294844,-75.93749999999999 43.06888777416965,-92.81250000000001 43.06888777416965,-92.81250000000001 29.53522956294844))', 4326)) = 1
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
| geom                                                                                                                                                                                                                                                                                                        | StyleIndex |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
|
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
SQLRowCount returns 1
1 rows fetched
SQL>
```

As you see StyleIndex is appearing as blank. But when I switch the order of StyleIndex and geom..

```

SQL> SELECT "StyleIndex", "Geometry" AS geom FROM (SELECT Geometry,LinkDir,StyleIndex,IsOneWay FROM MAPNIK WHERE SessionId = 69) as cdbq WHERE "Geometry".STIntersects(geometry::STGeomFromText('POLYGON((-92.81250000000001 29.53522956294844,-75.93749999999999 29.53522956294844,-75.93749999999999 43.06888777416965,-92.81250000000001 43.06888777416965,-92.81250000000001 29.53522956294844))', 4326)) = 1
+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| StyleIndex | geom                                                                                                                                                                                                                                                                                                        |
+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2          |
+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
SQLRowCount returns 1
1 rows fetched
SQL>
```

So I am attempted to go through all tds version 7.x to see which one displays accurately and seems 7.1 yields:

```
SELECT "Geometry" AS geom,"StyleIndex" FROM (SELECT Geometry,LinkDir,StyleIndex,IsOneWay FROM MAPNIK WHERE SessionId = 69) as cdbq WHERE "Geometry".STIntersects(geometry::STGeomFromText('POLYGON((-92.81250000000001 29.53522956294844,-75.93749999999999 29.53522956294844,-75.93749999999999 43.06888777416965,-92.81250000000001 43.06888777416965,-92.81250000000001 29.53522956294844))', 4326)) = 1
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
| geom                                                                                                                                                                                                                                                                                                        | StyleIndex |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
|                                                                                                                                                                                                                                                                                                             | 2          |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
SQLRowCount returns 1
1 rows fetched
```

Which seems to not cut/off the StyleIndex from the output. I recompiled with 7.1 and I am still having the same error where StyleIndex should be a 2 but Mapnik is reporting it a 0. Using the exact same mapnik configuration but with the mssql plugin `v0.1.48` and using windows it correctly reports 2 for styleindex.

Therefore I can only conclude that either my compiled version is wrong intermittently (returns geometry but not other fields correctly) or something about the plugin isn't playing nice to the newest freetds with unixodbc

**Comment by BHare1985** [comment]
I think I need to run the tests to see how whats going on, but its not obvious to me how to construct the makefile for it. I did it similarly to the plugin and while it compiled without error upon running it gives a segmentation fault. Using gdb on it yields nothing useful.

I tried copying the hello world test example in the make  file but I end up with `./test/run: error while loading shared libraries: libmapnik.so: cannot open shared object file: No such file or directory`

```
[root@ip-172-31-7-82 mapnik-mssql]# ldd mssql.input
        linux-vdso.so.1 =>  (0x00007ffc9386c000)
        libmapnik.so => not found
        libz.so.1 => /lib64/libz.so.1 (0x00007f0413219000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f0413015000)
        libodbc.so.2 => /usr/lib64/libodbc.so.2 (0x00007f0412dae000)
        libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f0412aaa000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f04127a8000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f0412592000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f0412376000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f0411fb5000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f041365a000)
        libltdl.so.7 => /usr/lib64/libltdl.so.7 (0x00007f0411dac000)
```

However that seems to be normal as suggested here: https://github.com/mapnik/node-mapnik/issues/234

Is there an example makefile to run the tests?

**Comment by gaspardle** [comment]
I'Il check it out and do some tests. I will also try to find the config that I used on linux and osx.

**Comment by BHare1985** [comment]
Thanks, and for the record this is what I used to get what I need for the plugin compiled for amazon linux

```
#!/bin/bash

yum install unixODBC unixODBC-devel
cd /tmp

wget ftp://ftp.freetds.org/pub/freetds/stable/freetds-patched.tar.gz
tar zxvf freetds-patched.tar.gz

cd freetds-*
./configure --prefix=/usr --sysconfdir=/etc --with-unixodbc=/usr --with-tdsver=7.3
make
make install

cat <<EOF >> /etc/odbc.ini
[connect]
Driver = SQL Server Native Client 11.0
Description = Microsoft SQL
Server = 68.43.177.158
Port = 1433
Database = TTDS
EOF

cat <<EOF >> /etc/odbcinst.ini
[SQL Server Native Client 11.0]
Description = FreeTDS unixODBC Driver
Driver      = /usr/lib/libtdsodbc.so
Setup       = /usr/lib/libtdsodbc.so
EOF

cat <<EOF >> /etc/odbcinst.ini
[SQLServerNativeClient11.0]
Description = FreeTDS unixODBC Driver
Driver      = /usr/lib/libtdsodbc.so
Setup       = /usr/lib/libtdsodbc.so
EOF

```

The DSN of `connect` is just an easy way to verify FreeTDS is installed correctly. You can test it using `osql -S connect -U <user> -P <password>` and it should output that it found the driver and was able to access it then it should connect and give you a SQL prompt if FreeTDS and unixOBDC are working correctly

Here is my compiled plugin: https://dl.dropboxusercontent.com/u/8518063/ShareX/2015/10/mssql.input

So just to reiterate, it gets the geometry and not the other fields. For example if I set mapnik to render CSS/styles on a field called StyleIndex , it will always return 0 for styleindex, but always return the geometry accurately. I tried various fields (ints) and all defaulted to 0 , even if I intercepted the query and ran it and saw the result was non-zero.

**Comment by BHare1985** [comment]
You are welcome to use the `68.43.177.158` server if you want. The DB name is TTDS and the user and pass are both ttds. You can find my mapnik xml here: https://gist.github.com/892d0f71dcaf7ba2ffaa

The `SELECT "StyleIndex", "Geometry" AS geom FROM (SELECT Geometry,LinkDir,StyleIndex,IsOneWay FROM MAPNIK WHERE SessionId = 69) as cdbq WHERE "Geometry".STIntersects(geometry::STGeomFromText('POLYGON((-92.81250000000001 29.53522956294844,-75.93749999999999 29.53522956294844,-75.93749999999999 43.06888777416965,-92.81250000000001 43.06888777416965,-92.81250000000001 29.53522956294844))', 4326)) = 1` query refers to a request for tile `5/8/12.png`  It should render this 
this image http://i.imgur.com/t8zmZkm.png  but in yellow (since StyleIndex is 2)...If its 0 then its grey.  

Oh and I verified with php-odbc that freetds is working as it should, the whole 7.1 vs 7.2 vs 7.3 is unrelated to this problem. I think the output of that debugger went whacky when it got a geometry binary back. So using 7.3 is preferred

**Comment by BHare1985** [comment]
Just an update, I was able to get rid of the `libmapnik.so: cannot open shared object file: No such file or directory` by following https://github.com/mapnik/mapnik/wiki/InstallationTroubleshooting  and adding a new file `/etc/ld.so.conf.d/whatever.conf` that contained the location of the Mapnik SDK's lib folder (I didn't install the lib folder to my `usr/local`which seems I should of.

I added this to the Makefile

```
test/run: $(BIN) test/main.cpp
    $(CXX) -o ./test/run test/main.cpp $(CXXFLAGS) $(LIBS)

test: test/run
    ./test/run
```

so running `make test` resulting in it saying `Problem loading plugin library: mssql.input (dlopen failed - plugin likely has an unsatisfied dependency or incompatible ABI)` which was really weird as I had fixed that problem and it doesnt appear with mapnik, then I just happen to stumble upon the solution which is to change `std::string mssql_plugin = "mssql.input";` to `std::string mssql_plugin = "./mssql.input";` ..weird that if it had issues finding it  before it didn't say anything about it.

Finally my tests were failing for `Data source name not found, and no default
  driver specified` but the connection string was correct. Turns out that my freeTDS/unixODBC requires driver to be in all caps so `driver=` had to be turned into `DRIVER=` and then it worked.

I modified the tests to use `CHECK` instead of `REQUIRED`, as a required will stop other tests from running. See here: https://gist.github.com/7d5302751ded22016206

Results of the tests:

```
run is a Catch v1.1 b1 host application.
Run with -? for options

-------------------------------------------------------------------------------
mssql
  create datasource
  query features
-------------------------------------------------------------------------------
test/main.cpp:13
...............................................................................

test/main.cpp:75: FAILED:
  CHECK( 2147483648 == _bigint_value )
with expansion:
  2147483648 == 0

test/main.cpp:80: FAILED:
  CHECK( 1 == int_value )
with expansion:
  1 == 0

test/main.cpp:85: FAILED:
  CHECK( 1.25 == Approx(float_value) )
with expansion:
  1.25 == Approx( 0.0 )

test/main.cpp:90: FAILED:
  CHECK( 1.25 == Approx(real_value) )
with expansion:
  1.25 == Approx( 0.0 )

test/main.cpp:95: FAILED:
  CHECK( true == bit_value )
with expansion:
  true == false

test/main.cpp:116: FAILED:
  CHECK( 1.25 == Approx(money_value) )
with expansion:
  1.25 == Approx( 0.0 )

===============================================================================
test cases:  1 |  0 passed | 1 failed
assertions: 24 | 18 passed | 6 failed


```

So clearly it is what I expected, for some reason it is getting zero for all values besides geometry.

**Comment by BHare1985** [comment]
I compiled with gdb flag and been debugging the testcases, I have a breakpoint on result.hpp function `getBigInt` and I get a retcode of 100 (SQL_NO_DATA) : http://i.imgur.com/2nnHHzZ.png which eventually does attempt to trigger a MAPNIK_LOG_WARN but by default that is never turned on so I added

```
#include <mapnik/debug.hpp>
...
mapnik::logger::instance().set_severity(mapnik::logger::debug);
```

To the test and I am not getting any warningns in STDOUT, but I think debug/warnings should be visible with the test to help debug.

**Comment by BHare1985** [comment]
Hey, great work on this. I just recompiled and ran the test cases and they all passed. I am not sure why you reopened the issue but as far as I am concerned, this issue is fixed.

Do note that the filesize went from 220kb to 3365kb

**Comment by gaspardle** [comment]
I just wanted to be sure that it was working on your side. 

There was some differences on how freetds and the microsoft driver behave, I did not benchmark the code to see if there's any performances issue.
